man/fi_trigger.3.md

```diff
@@ -94,6 +103,124 @@ struct fi_trigger_threshold {
+	} value;
+};
+
+struct fi_trigger_xpu {
+	int count;
+	enum fi_hmem_iface iface;
+	union {
+		uint64_t reserved;
+		int cuda;
+		int ze;
+	} device;
+	struct fi_trigger_var *var;
+};
+```
+
+On input to a triggered operation, the iface field indicates the software
+interface that will be used to write the variables.  The device union
+specifies the device identifier.  For valid iface and device values, see
+[`fi_mr`(3)](fi_mr.3.html).  The iface and device must match with the
+iface and device of any local HMEM memory regions.  Count should be set
+to the number of fi_trigger_var structures available, with the var field
+pointing to an array of struct fi_trigger_var.  The user is responsible for
+ensuring that there are sufficient fi_trigger_var structures available and of
+an appropriate size.  The count and size of fi_trigger_var structures
+can be obtained by calling fi_getopt() on the endpoint with the
+FI_OPT_XPU_TRIGGER option.  See [`fi_endpoint`(3)](fi_endpoint.3.html)
+for details.
+
+Each fi_trigger_var structure referenced should have the datatype
+and count fields initialized to the number of values referenced by the
+struct fi_trigger_val.  If the count is 1, one of the val fields will be used
+to return the necessary data (val8, val16, etc.).  If count > 1, the data
+field will return all necessary data used to signal the trigger.  The data
+field must reference a buffer large enough to hold the returned bytes.
+
+On output, the provider will set the fi_trigger_xpu count to the number of
+fi_trigger_var variables that must be signaled.  Count will be less than or
+equal to the input value.  The provider will initialize each valid
+fi_trigger_var entry with information needed to signal the trigger.  The
+datatype indicates the size of the data that must be written.  Valid datatype
+values are FI_UINT8, FI_UINT16, FI_UINT32, and FI_UINT64.  For signal
+variables <= 64 bits, the count field will be 1.  If a trigger requires writing
+more than 64-bits, the datatype field will be set to FI_UINT8, with count set
+to the number of bytes that must be written.  The data that must be written
+to signal the start of an operation is returned through either the value
+union val fields or data array.
+
+Users signal the start of a transfer by writing the returned data to the
+given memory address.  The write must occur from the specified input XPU
+location (based on the iface and device fields).  If a transfer cannot
+be initiated for some reason, such as an error occurring before the
+transfer can start, the triggered operation should
+be canceled to release any allocated resources.  If multiple variables are
+specified, they must be updated in order.
+
+Note that the provider will not modify the fi_trigger_xpu or fi_trigger_var
+structures after returning from the data transfer call.
+
+In order to support multiple provider implementations, users should trigger
+data transfer operations in the same order that they are queued and should
```

Comment:
Could provide as query output whether triggers may be processed out of order by the provider.